packs/democratic_csi_nfs/README.md

```diff
@@ -0,0 +1,180 @@
+# `democratic-csi` CSI plugin
+
+This pack deploys two jobs that run the
+[`democratic-csi`](https://github.com/democratic-csi/democratic-csi)
+CSI plugin. The node plugin tasks will be run as a system job, and the
+controller tasks will be run as a service job.
+
+## Client Requirements
+
+This pack can only be run on Nomad clients that have enabled volumes and
+privileged mode for the Docker task driver. In addition, clients will
+need to have a source NFS volume mounted to any client host that runs
+the controller task.
+
+### Example NFS Server
+
+The following is an example of installing and configuring a NFS server
+on an apt-based Linux distrobution (ex. Debian or Ubuntu). This
+configuration exports the directory `/var/nfs/general` to any Nomad
+client on the `192.168.56.0/24` address space (this is commonly used
+for Vagrant hosts on Virtualbox).
+
+```sh
+sudo apt-get install nfs-kernel-server
+sudo mkdir /var/nfs/general
+
+sudo cat <<EOF > /etc/exports
+/var/nfs/general 192.168.56.0/24(rw,sync,no_subtree_check,no_root_squash)
+EOF
+
+sudo systemctl enable nfs-kernel-server
+sudo systemctl start nfs-kernel-server
+```
+
+The `democratic-csi` controller is unusual in needing to mount the NFS
+volume because NFS doesn't have a remote API other than filesystem
+operations. So this plugin has to have the NFS export bind-mounted
+into the plugin container. Any client running the controller will need
+to have the NFS export in the host's `/etc/fstab`. The following
+configuration is an example for the NFS export shown above, assuming
+that the NFS server can be found at IP `192.168.56.60`. The mount
+point `/srv/nfs_data` shown here should be used for the
+`nfs_controller_mount_path` variable.
+
+```
+sudo cat <<EOF >> /etc/fstab
+192.168.56.60:/var/nfs/general /srv/nfs_data nfs4 rw,relatime 0 0
+EOF
+
+```
+
+## Variables
+
+The following variables are required:
```

Comment:
Is there a way to express this requirement in our `variables.hcl`? There aren't really reasonable defaults for these values here.

Reply:
No way :(

Reply:
If you don't default them, they should end up being required? If they aren't this seems like a bug. (It's inconsistent with my experience with Nomad's HCL2, that is)

Reply:
No, unfortunately. For example if I render this pack with no variables set I end up with a partially empty template:
 
```
      template {
        destination = "${NOMAD_TASK_DIR}/driver-config-file.yaml"

        data = <<EOH
driver: nfs-client
instance_id:
nfs:
  shareHost:
  shareBasePath: ""
  controllerBasePath: "/storage"
  dirPermissionsMode: "0777"
  dirPermissionsUser: root
  dirPermissionsGroup: root
EOH
      }
```

For Nomad fields, this usually up throwing an error when we try to submit the job because Nomad will reject it. But for something like the interior of a template, Nomad has no way of knowing whether the field is valid.